### PR TITLE
OpenTelemetry Collector Ecosystem Integration

### PR DESCRIPTION
**Pull request by Nicolas-MacBeth**
This PR is meant to define the folder and introduce the community to the receiver I've been working on, and get feedback on it. It includes a `README` with the current state of the config which is also a really good in-depth config explanation.

Hi! I'm Nic, an intern @ Google. 

`prometheus_exec` is a receiver meant to make it easy for users to get up-and-running quickly, collecting metrics from their favorite third-party services (that don't normally support instrumentation). This frictionless experience should help with adoption and be an attractive solution for those holding out on using the Collector because their systems/services were not compatible. In a nutshell, it runs binaries of [Prometheus exporters](https://prometheus.io/docs/instrumenting/exporters/) (custom-built code that make third-party services, such as MySQL, Redis, etc. communicate the Prometheus protocol) and spawns the equivalent Prometheus receiver to collect the metrics. It handles crashing binaries by applying exponential backoff, can assign a random port and supports string templating. Please read the `README` (it really has a lot of useful info) or ask me for more details :) !

README link with markdown applied (**much** nicer read): [link](https://github.com/Nicolas-MacBeth/opentelemetry-collector-contrib/blob/readme_and_structure/receiver/prometheusexec/README.md)

For example, a user wants to quickly get metrics from their MySQL and Apache servers running on a VM. Simply, they indicate a couple of things in the configuration file, download the exporter binaries and run the Collector. It will handle starting the exporters, starting the equivalent receivers and making everything work like magic (hopefully).

Example configuration:
```yaml
receivers:
    prometheus_exec/apache:
        exec: ./apache_exporter --retry=false
        port: 1234 
        scrape_interval: 60s

    prometheus_exec/postgresql:
        exec: ./postgresql_exporter --port {{port}} --config conf.xml
        port: 9876
        env:
          - name: DATA_SOURCE_NAME
            value: user:password@url:port/dbname
```

### CODE REVIEW

File: receiver/prometheusexec/README.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+// TODO: Remove above message when done
+
+# Prometheus_exec Receiver
+
+This receiver makes it easy for a user to collect metrics with the Collector from third-party services such as MySQL, Apache, Redis, etc. It's meant for people who want a plug-and-play solution to getting metrics from those third-party services that don't natively export metrics or speak any instrumentation protocols. Through the configuration file, you can indicate which binaries to run (usually [Prometheus exporters](https://prometheus.io/docs/instrumenting/exporters/), which are custom binaries that will expose the services' metrics to the Prometheus protocol). It supports starting binaries with flags and environment variables, retrying them with exponentional backoff if they crash, string templating, and random port assignments. If you do not need to spawn the binaries locally, please condider using the [core Prometheus receiver](https://github.com/open-telemetry/opentelemetry-collector/tree/master/receiver/prometheusreceiver) or the [Simple Prometheus receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/receiver/simpleprometheusreceiver).
```
Comment (james-bebbington) [inline receiver/prometheusexec/README.md:6]: I have a little bit of trouble reconciling that opening sentence. As someone reading this README for this first time, I would get very confused by this, as, for example, there is already a Redis receiver in this same repo, i.e. we are talking about "Redis -> Prom Exporter -> Prom Receiver -> Export" vs "Redis -> Redis Receiver -> Export"

I think it would be clearer to open with something like "collecting metrics **via third party applications that expose prometheus metrics**" (or maybe just "collecting metrics via prometheus exporter plugins, or similar")?

And due to the multiple different ways you can generate metrics from applications, it may be worth adding a line that justifies why this is useful - e.g. "to make use of the entire existing ecosystem of Prometheus exporter plugins" or something like that.

File: receiver/prometheusexec/README.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+## Overview
+Here's an example Collector configuration file that starts two `prometheus_exec` receivers and two exporters.
```
Comment (james-bebbington) [inline receiver/prometheusexec/README.md:9]: I think you can probably limit this README to just focus on the `prometheus_exec` config. The rest of the config is already well documented in the general Collector docs.

File: receiver/prometheusexec/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+service:
+    pipelines:
+        metrics:
+            receivers: [prometheus_exec/mysqld, prometheus_exec/postgresql]
+            exporters: [stackdriver, file]
+```
+
+## Config
+For every `prometheus_exec` defined in the configuration file, a command will be run (you *should* start a binary for this receiver to serve its purpose) and an equivalent Prometheus receiver will be instantiated to scrape its metrics.
```
Comment (james-bebbington) [inline receiver/prometheusexec/README.md:41]:
```suggestion
For each `prometheus_exec` defined in the configuration file, the specified command will be run. This command *should* start a binary that exposes Prometheus metrics on the specified port. A Prometheus receiver will be instantiated to scrape this endpoint.
```

File: receiver/prometheusexec/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+- ### prometheus_exec (required)
+`prometheus_exec` receivers should hierarchically be placed under the `receivers` key. You can define as many of these as you want, and they should be named as follows: `prometheus_exec/custom_name`. The `custom_name` should be unique per receiver you define, and is important for logging and error tracing. If no `custom_name` is given (i.e. simply having `prometheus_exec`), one will be generated for you by using the first word in the command to be executed (doing this limits the amount of receivers to 1 though since you can't define multiple receivers with duplicate names, so it's not recommended). Example:
```
Comment (james-bebbington) [inline receiver/prometheusexec/README.md:44]: Isn't the first part (named receivers) just regular Collector behaviour? And why bother with the second part at all? Is this because you always need to create a named Prometheus receiver internally? To keep this simpler, perhaps just make the name mandatory for this receiver then?

File: receiver/prometheusexec/README.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+2. **All** instances of `{{port}}` in any string of any key for the enclosing `prometheus_exec` will be replaced with either the port value indicated or the randomly generated one if no port value is set with the `port` key.
+
+Example:
+
+```yaml
+receivers:
+    prometheus_exec/apache:
+        exec: ./apache_exporter --retry=false
+        port: 1234
```
Comment (james-bebbington) [inline receiver/prometheusexec/README.md:80]: Can you make sure all the examples in this file are legitimate as people may copy/paste, e.g. looks like the default port for this exporter is [9117](https://github.com/Lusitaniae/apache_exporter/blob/8c2e8538a06c63ed247001fad9cd3a63d97ced97/apache_exporter.go#L28)

**Review by james-bebbington — COMMENTED** [review]
Left a few comments

---

Side note:

> README link with markdown applied (**much** nicer read): [link](https://github.com/Nicolas-MacBeth/opentelemetry-collector-contrib/blob/readme_and_structure/receiver/prometheusexec/README.md)

You can get to the raw file from the PR via:

image

File: receiver/prometheusexec/README.md (hunk 2 above)
Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:9]: Okay, I'll remove! thanks

File: receiver/prometheusexec/README.md (hunk 1 above)
Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:6]: Thanks, I fixed the wording for multiple sentences in this paragraph!

File: receiver/prometheusexec/README.md (hunk 3 above)
Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:41]: Thanks for the suggestions. I kept most of what you wrote, but removed the part about ports and such since there's multiple ways for the user to go about choosing a port!

File: receiver/prometheusexec/README.md (hunk 4 above)
Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:44]: Not sure what you mean by first part here, but yes, the Collector normally works by having 1 receiver per receiver defined, however I wanted to make it clear since it's different from the "original" Prom receiver since that one works with 1 receiver having multiple `scrape_configs`.

The `custom_name` is mostly to prefix logs from the processes running, to help the user debug or read logs - they'll know which process is spitting out those logs. I just didn't want to make it mandatory in case the user forgets or something. I'm open to making it required, though. I thought just generating one for the user would be nice if they forgot or something!

File: receiver/prometheusexec/README.md (hunk 5 above)
Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:80]: Good point - I'll make them as real as possible! Thanks

Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:80]: Good point, I'll make them as real as possible - thanks!

File: receiver/prometheusexec/README.md (hunk 4 above)
Comment (james-bebbington) [reply receiver/prometheusexec/README.md:18]: Sorry by first part I meant the first three sentences.

> different from the "original" Prom receiver since that one works with 1 receiver having multiple scrape_configs.

That's not strictly true. Users can create multiple named promethus receivers if they like.

I think you may be re-implementing something that the Collector already supports out of the box (other tha ntaking the name from the command by default). It might not be as well documented as I thought. The [design docs](https://github.com/open-telemetry/opentelemetry-collector/blob/master/docs/design.md#receivers) link to this [Google Doc](https://docs.google.com/document/d/1NeheFG7DmcUYo_h2vLtNRlia9x5wOJMlV4QKEK05FhQ/edit#heading=h.74px12cea3py) which specifies how to named components can be created (search for "type[/name]" in that doc)

Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:18]: Oh I see what you mean! In the configuration, I'm just trying to indicate to the user that that's how **this** receiver should be invoked (separate receivers with unique names), which differentiates from how Prometheus receivers are **usually** defined (I understand you can have multiple receivers but the *usual* way seems to be simply multiple `scrape_configs`). To be clear, I'm not re-implementing the `receiver[/custom_name]` structure (I'm using what's already built into the Collector), only stating its necessity here! Also, I'm letting the user know where/how the `custom_name` will be used (logging purposes for now).

I think we're agreeing here: since multiple separate receivers need to be defined to have multiple `prometheus_exec`s running at once and the Collector already enforces unique names for multiple receivers, I'm just letting the user know how they should write their config file, and I'm not re-implementing that!

Does this clear any confusion? Thanks for the input btw, very appreciated. 

- You think I should re-word the this part of the README to include the fact that they need custom names if they have multiple receivers and that that is normal Collector behavior, and then say that this specific receiver uses the `custom_name` for logging?

Comment (james-bebbington) [reply receiver/prometheusexec/README.md:18]: I'm still a little bit confused. If this just uses the default behaviour, then how are you implementing this part? "If no `custom_name` is given (i.e. simply having `prometheus_exec`), one will be generated for you by using the first word in the command to be executed"

Are you planning on overriding the Name by calling `NamedEntity.SetName()` from within the receiver? I'd probably recommend not doing that as that then deviates from how all other components behave.

I think I'd probably remove this entire section, or maybe change it to:

"To run multiple Prometheus exporters, create multiple named `prometheus_exec` receivers: (yaml example)"

Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:18]: Oh I see where the confusion is! `custom_name` here is simply used by my receiver as a prefix for logging. Whether the user puts something for `name` in `prometheus_exec[/name]` or not, I won't do anything about that. I won't be changing the internal name the Collector uses for that receiver, I'll simply be generating one for **my** use for this receiver.

To recap, if the user defines a name in `prometheus_exec[/name]`, then I use that as my `custom_name` (used only by my receiver for logging prefixes), and if not I generate one for them.

Do you still think I should remove it?

Comment (james-bebbington) [reply receiver/prometheusexec/README.md:18]: So it sounds like you just want to be able to distinguish between logs from different named receivers? If you think about it, functionality where you include the name, type, etc of the component in the logs would be much more useful if implemented as generic behaviour of the collector. Otherwise, it would need to be implemented for each component separately!

And, it is already implemented as generic behaviour. As long as you use the `logger` passed in to your component via the `factory.CreateMetricsReceiver` (etc) function, that logger has already been configured to log that information. See: https://github.com/open-telemetry/opentelemetry-collector/blob/master/service/builder/pipelines_builder.go#L159

If you check the logs you should see something like this:
```
2020-07-04T13:59:15.553+1000	INFO	receiver/my_receiver.go:213	My log message.	{"component_kind": "receiver", "component_type": "my_receiver_type", "component_name": "my_receiver_name", "datatype": "metrics"}
```

Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:18]: I might not be able to use the `logger` for logs coming from the subprocesses, but thanks for the info! I'll look into it more and update the README accordingly. Thanks for the discussion! :)

Comment (Nicolas-MacBeth) [reply receiver/prometheusexec/README.md:18]: Update: I removed this part of the README and removed the name generation code. As per normal Collector behavior, the user needs unique `custom_names` if they want multiple of the same receiver running at once. This way, a `custom_name` is very likely to be defined by the user. On the off chance that the user only defines one of this receiver **and** does not give it a `custom_name`, the logs won't have a prefix but that's not a problem since only one process will be logging, which makes it easy to know which process it's coming from.

To recap, this part of the README is being removed and the `custom_name` will be defined by the user per normal Collector behavior, no `custom_name` generation here.

**Comment by james-bebbington** [comment]
@Nicolas-MacBeth Close this?

**Comment by bogdandrutu** [comment]
@james-bebbington why closing? Is it no longer needed?

**Comment by Nicolas-MacBeth** [comment]
It is needed @bogdandrutu ! I had just done it for early review and tweaking of the design/README. The full PR is done and can be found here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/453. That PR was a little massive, so I've broken it into 2 chunks, first part is here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/499. As soon as the first part is merged, I'll make the second.

To recap, I'm closing this PR, have a second huge PR with the entire receiver and finally have a third PR with a chunk of it (easier to review and merge). Thanks @james-bebbington

**Comment by james-bebbington** [comment]
If this README is still up to date, we can keep this PR - it says under development. I just noticed you had the same file committed in both PRs.

**Comment by Nicolas-MacBeth** [comment]
I'll close it since the README is out of date, and to not have too many PRs open at once!
